### ISSUE TITLE
Questions about S, convergence, usage on unseen data, update of regularization methods

### ISSUE THREAD

**Issue by grimmdaniel** [zc8340311/RobustAutoencoder#13]
Hi Chong,

First of all, thank you for your interesting article. We tried to implement your idea in Pytorch, and use it in production for anomaly detection. In our case we have a huge dataset of measurements without labels or ground truth. We know that the data has some outliers which we could not catch with linear models. So we would like to train or model (AE) to learn on non anomalous data if possible. However, we ran into some issues, and I would like to ask some questions, if you don't mind. I'm trying to use the outlier detection model with l21 regularization.

So the idea is to separate anomalous data into the S matrix. In my understanding it is only possible to use it for the training data to make prediction of anomalous instances, right? What do you recommend for classifying on unseen data? Normally autoencoders are used in a way, that the reconstruction error tells how much an instance is an outlier. In this case, what is the difference between relying on the S matrix instead of the reconstruction error? Should be there any correlation between these? (like high reconstruction error <-> S is non zero in that row)

In the article you stated that "non zero rows are outliers" in S. Does it mean if at least one element in a row of S is greater than zero, then the whole row(data instance) is an anomaly? Or the whole row has to be > 0 to consider it as an outlier?
In my case I cannot choose a correct lambda value which would give at least reasonable results. The problem is that always the same columns have non zero values. If I make lambda larger I will have more zeros in S but still all non zero elements will be in the same columns, only difference is that some additional columns will become 0. This means I will never get totally zero rows only if the whole matrix is zero (a little smaller lamdba would still give at least on column full of non zero elements)

As I understood, lamdba depends on the actual input data, in my case it ranges around 10-140. (Above 140 the whole S is 0, below 0-10 S does not have any 0.) 
What seems strange is that in your case the plotted loss function decreases gradually with steps, every time you minimalize the L,S matrices, but in my case right after the first model training (so basically the first time L-S happens when S is not the zero matrix), the loss drops drastically to very small values (like 0.001) then the model hardly improves, sometimes the second L-S even makes the model diverge.

Also, I've made some modifications on the shrink methods, to be able to use them on high amount of input data. Basically, I've changed them to vectorized functions:

```python
def l1shrink(mat, eps):
    output = np.zeros_like(mat)
    cond_pos = mat > eps
    cond_neg = mat < -eps
    output[cond_pos] = mat[cond_pos] - eps
    output[cond_neg] = mat[cond_neg] + eps
    return output
```

```python
def l21shrink(mat, eps):
    output = np.zeros_like(mat)
    norm = np.linalg.norm(mat, ord=2, axis=0)
    condition = norm > eps
    output[:, condition] = mat[:, condition] - eps * mat[:, condition] / norm[condition]
    return output
```

Thank you,
Daniel
